lib/uri/common.rb

```diff
@@ -518,22 +518,56 @@ def self.encode_www_form(enum, enc=nil)
     end.join('&')
   end
 
-  # Decodes URL-encoded form data from given +str+.
+  # Returns name/value pairs derived from the given string +str+.
   #
-  # This decodes application/x-www-form-urlencoded data
-  # and returns an array of key-value arrays.
+  # The method may be used to decode the body of Net::HTTPResponse object +res+
+  # for which <tt>res['Content-Type']</tt> is <tt>'application/x-www-form-urlencoded'</tt>.
   #
-  # This refers http://url.spec.whatwg.org/#concept-urlencoded-parser,
-  # so this supports only &-separator, and doesn't support ;-separator.
+  # The returned data is an array of 2-element subarrays;
+  # each subarray is a name/value pair (both are strings).
   #
-  #    ary = URI.decode_www_form("a=1&a=2&b=3")
-  #    ary                   #=> [['a', '1'], ['a', '2'], ['b', '3']]
-  #    ary.assoc('a').last   #=> '1'
-  #    ary.assoc('b').last   #=> '3'
-  #    ary.rassoc('a').last  #=> '2'
-  #    Hash[ary]             #=> {"a"=>"2", "b"=>"3"}
+  # A simple example:
+  #
+  #   URI.decode_www_form('foo=0&bar=1&baz')
+  #   # => [["foo", "0"], ["bar", "1"], ["baz", ""]]
+  #
+  # The returned strings have certain conversions,
+  # similar to those performed in URI.decode_www_form_component:
+  #
+  #   URI.decode_www_form('f%23o=%2F&b-r=%24&b+z=%40')
+  #   # => [["f#o", "/"], ["b-r", "$"], ["b z", "@"]]
+  #
+  # The given string may contain consecutive separators:
+  #
+  #   URI.decode_www_form('foo=0&&bar=1&&baz=2')
+  #   # => [["foo", "0"], ["", ""], ["bar", "1"], ["", ""], ["baz", "2"]]
+  #
+  # A different separator may be specified:
+  #
+  #   URI.decode_www_form('foo=0--bar=1--baz', separator: '--')
+  #   # => [["foo", "0"], ["bar", "1"], ["baz", ""]]
+  #
+  # Further details:
+  #
+  # - The given +str+ is first split into the substrings returned by:
+  #
+  #     str.b.each_line(separator)
+  #
+  #   See {String#b}[https://docs.ruby-lang.org/en/master/String.html#method-i-b]
+  #   and {String#each_line}[https://docs.ruby-lang.org/en/master/String.html#method-i-each_line].
+  #
+  # - Each substring +string+ is chomped,
+  #   and split into a name/value pair by:
+  #
+  #     string.partition('=')
+  #
+  #   See {String#chomp!}[https://docs.ruby-lang.org/en/master/String.html#method-i-chomp-21]
+  #   and {String#partition}[https://docs.ruby-lang.org/en/master/String.html#method-i-partition].
+  #
+  # - Each name and value is force-encoded to encoding +enc+, then scrubbed.
+  #   See {String#force_encoding}[https://docs.ruby-lang.org/en/master/String.html#method-i-force_encoding]
+  #   and {String#scrub!}[https://docs.ruby-lang.org/en/master/String.html#method-i-scrub-21].
```

Comment:
I don't think it's a good idea to talk about the implementation in the documentation.

I think there's two things here worth nothing:

1. Argument `str` must be an ASCII string. If it isn't, then an ArgumentError will be raised:
```ruby
p URI.decode_www_form("foo=0&bar=1&baz\u{6666}")
# raises ArgumentError
```

2. The strings in the array returned will have their encodings set to `enc` and all invalid characters removed (using `String#scrub`).